### ISSUE TITLE
External CFE message definitions should not depend on values from cfe_platform_cfg.h or osconfig.h

### ISSUE THREAD

**Issue by skliper** [nasa/cFE#43]
Some structure definitions that define specific ground commands and telemetry messages use array sizes for strings that are either defined in osconfig.h (e.g. OS_MAX_PATH_LEN) or cfe_platform_cfg.h (e.g. CFE_ES_MAX_APPLICATIONS).  

However, by definition, these header files contain parameters that are tunable to the specific CPU.  For instance, a resource-constrained CPU may need to reduce CFE_ES_MAX_APPLICATIONS to a small number to save memory.  It is therefore problematic to use these values to define messages that serve as a data definition across processors, particularly in a multi-board setup.

External message definitions must be limited to **mission-wide** configuration parameters only, such as cfe_mission_cfg.h, where all CPUs as well as the ground system will have the same values.

**Comment by skliper** [comment]
_Imported from trac issue 12.  Created by jphickey on 2014-12-30T21:26:08, last modified: 2019-03-05T14:58:28_

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-05-04 10:11:10:_

Will resolve this as part of next change set.

As a related note, {{{cfe_platform_cfg.h}}} __should not__ be available for inclusion by applications.  The values defined within this file are tuning parameters for the CFE core, and should not be used by applications intended to be portable.  This keeps the build simpler, and also allows application binaries to be shared between multiple CPU builds that use the same architecture (i.e. the same binary for application "z" can be loaded onto cpu1 and cpu2 if they both use the same CPU architecture).

Most applications using this header are easily fixed:
- Some apps were including it but not using it for anything - so removing it "just works" (probably originally a cut and paste)
- The application can be modified to not require the actual value
- The value of the parameter can be acquired at run time rather than compile time
- The value can become a mission-level compile-time tunable for this application

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-05-11 10:00:18:_

Pushed commit 2ad6078 for review

**EDIT** 5/29/15 -- Per feedback - holding off on this commit until a later date. Please see below for updated commit ID for review.

**Comment by skliper** [comment]
_Trac comment by sduran on 2015-05-15 15:07:13:_

Changes look ok for cfe code, but there could be impacts to CFS apps, not to mention existing user apps.  Probably need an apps ticket to assess CFS apps for impacts and update.

**Comment by skliper** [comment]
_Trac comment by acudmore on 2015-05-18 15:56:38:_

I agree with Steve in the comment above. We need to determine the impacts. 

I do like the idea of binary compatible apps, and it is a good idea to not define telemetry based on processor specific defines. 

We may end up having to split osconfig.h into mission defines for path, file name lengths, and such. The sizes of the OS data structures (such as how many tasks, queues ) could go in a processor config file. We would still have to be careful with the mission wide file system defines, because there still could be a case for different maximum filename lengths on two different processors.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-05-18 16:26:27:_

Replying to [comment:4 sduran]:
> Changes look ok for cfe code, but there could be impacts to CFS apps, not to mention existing user apps.  Probably need an apps ticket to assess CFS apps for impacts and update.

Replying to [comment:5 acudmore]:
> I agree with Steve in the comment above. We need to determine the impacts. 

Yes - this change is (unfortunately) not transparent to the apps that are currently including platform-specific files.  I have updated all of the CFS applications so they can all be made to work, but there were some minor modifications to do on some of them.

**Comment by skliper** [comment]
_Trac comment by jwilmot on 2015-05-19 10:44:10:_

Although I aggree with the rationale, I'm not sure this meets the "ROI" test. The change potentially impact quite a few files and the memory savings is likely small. Maybe add notes in the app developers guide to deprecate app use of cfe_platform_cfg.h?

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-05-21 15:35:19:_

Reworking this change to address some of the issues and feedback

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-05-29 09:49:22:_

This is ready for re-review, as commit [changeset:beea9d3]

The new branch name is "trac-12-separate-api-platform-definitions"

Per feedback - I have pulled out the changes regarding inclusion of {{{cfe_platform_cfg.h}}}.  This can be deferred to a later date, but I would still like to get in the main focus of this trac which is to separate the size definitions within API/Message structures from the platform config.

The new commit focuses specifically on this aspect of separating the message/API binary formats from the platform configuration, but leaves the inclusion paths as-is.  This is an important prerequisite of EDS support but it is relevant/important even if EDS is not used.

**DEPENDENCY NOTE**: because strings in messages may be a different size than what the local platform/OS defines, this //requires// the fix for #86 to be included as well in order to guarantee things continue to work as expected.  The commit ID listed above is a descendant of the trac #86 fix for this reason.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-05-29 09:55:37:_

Replying to [comment:7 jwilmot]:
> Although I aggree with the rationale, I'm not sure this meets the "ROI" test. The change potentially impact quite a few files and the memory savings is likely small. Maybe add notes in the app developers guide to deprecate app use of cfe_platform_cfg.h?

Additional note - the new commit I just pushed will __not__ affect the ability to include the platform config and therefore does not affect compatibility with existing applications like the previous commit would have.  The new commit accomplishes the main goal of what I needed to do here for EDS support.

I still want to deprecate usage of "cfe_platform_cfg.h" from applications, but this can be deferred to a later build.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-07-15 13:54:30:_

== REBASED COMMIT ==

This has been rebased to the {{{development}}} branch approved at the 2015-07-14 CCB.

Rebased commit ID: [changeset:b8f8385]

Note that this still depends on #86 -- as of this writing #86 is not yet included in a development merge.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-07-27 15:00:00:_

Approve.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-07-27 15:21:35:_

**Note to reviewers:**

Although this does not directly prevent apps from including `cfe_platform_cfg.h`, this does change the name of certain definitions within that file and as such it does affect compatibility with existing applications.

The name changes are intentional - to give an example, application code might be referencing `CFE_ES_MAX_APPLICATIONS`.  

Depending on the context:
* It may be using this macro to interact with the local CFE ES instance.  In general this code should be rewritten to use the public function call API **only** and not rely on such macros at all.  However in the meantime this could be changed to  `CFE_ES_PLATFORM_APPLICATIONS` to preserve existing behavior.
* It may be using this macro to interpret software bus packets to/from CFE ES. In this case the correct value should be `CFE_ES_MSG_MAX_APPLICATIONS`

This means that neither definition is "valid" for all use cases.  By //not// preserving the existing name, code that references this value will trigger an error and force some examination to see what context this is being used in so the correct modification can be done.

I have written a corresponding ticket and fix for any issues for all existing cFS applications and those are in the [cfs_apps:] repository.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-08-11 13:55:44:_

Latest commit: [changeset:11cbd8c]

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-08-21 18:06:23:_

Please identify the ticket numbers in the cfs_apps repository.

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-08-24 18:06:39:_

How is the EDS work affected if this change set is NOT applied?

**Comment by skliper** [comment]
_Trac comment by sstrege on 2015-08-24 18:08:56:_

From Joe's email:

It was discussed during the last CCB that the applications would all need some sort of update in order to work with the changes introduced by CFE ticket #43.  Below is a list of tickets that allow all the apps to build using the platform configurations modified as part of CFE ticket #43.

Note - these also fix some other compiler warnings in addition to adapting to the platform config changes.

Also note - not proposing that we necessarily review these in detail at the upcoming CCB, as we probably do not have enough time.  I would rather go through the rest of the tickets on the previous list, possibly deferring #43 again until the 9/8 CCB if necessary until these changes can be reviewed.  The main objective here is to give an idea of the type of fixes that were required to apps in order to correctly build with the modified platform configuration.

#56: Fix MD using CFE_TBL definitions
Branch trac-25-md-sendevent-fix / commit 72d41b7

#57:  Fix HS build errors with strict settings Branch trac-26-fix-hs-build-errors / commit 402591f

#58: Fix HK build with strict settings
Branch trac-27-fix-hk-build-errors / commit 6a940b9

#60: Fix DS verification build error
Branch trac-29-fix-ds-build-error / commit 70e309a

#61: Fix CS build error
Branch trac-30-fix-cs-build-error / commit 4a10d1f

#65: Fix CF build errors
Branch trac-34-fix-cf-build-errors / commit 94f3d50
(#65 contains additional fixes beyond just the platform config related items - may split this up into multiple tickets if necessary)

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-08-31 11:43:18:_

== UPDATED ==

This was discussed at the 2015-08-25 CCB meeting and it was agreed to produce a "reduced" version of this change set that did not change any of the names within `cfe_platform_cfg.h`.

This is available as commit [changeset:b153c13], branch name `trac-12-platform-msg-limits-take2`

This cuts the number of affected files from 60 (original commit) to 18.

The remaining changes focus on the following areas:
* the `cfe_mission_cfg.h` now contains mission-scope size definitions telemetry/command objects.
* message payload definitions within in the `.h` files are updated to use the mission-scope definitions.
* In places where the messages are sent or received, the code is also adjusted accordingly.  When possible, the `sizeof()` operator is preferred for this operation so the code does not assume a certain size for the I/O structures.  Additionally, the code is updated to allow different size limits (padding out the I/O structure if it is larger in size than the local platform limit).

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-09-08 10:04:48:_

Related: Added ticket #142 regarding the naming convention for mission scope vs. platform scope macros.

**Comment by skliper** [comment]
_Trac comment by glimes on 2015-09-08 10:16:18:_

Approve.

I'll need to make sure to get all of the right headers in place in the sample_defs
area in the cfs_test tree when this goes live. This is all worked out correctly now
in the `ic-ccb-review` branch.

**Comment by skliper** [comment]
_Trac comment by cdknight on 2015-09-08 12:15:48:_

Replying to [comment:1 jphickey]:
>...{{{cfe_platform_cfg.h}}} __should not__ be available for inclusion by applications....
> - The value of the parameter can be acquired at run time rather than compile time

I'd like to retain as much visibility of compile/runtime configuration as possible and require as little re-writing of app code as possible. Global symbols could be used to give everyone access, but will increase the heap size which may be a problem for low-memory environments. Also, adds a memory read to every access. An array could be used to reduce the symbol table impact.

A simple example:

cfe_platform_cfg.h:
{{{
...
extern const uint32 CFE_SB_DEFAULT_MSG_LIMIT;
extern const uint32 CFE_SB_BUF_MEMORY_BYTES;
...
extern const char * const CFE_SB_DEFAULT_ROUTING_FILENAME;
...
}}}
cfe_platform_cfg.c:
{{{
...
const uint32 CFE_SB_DEFAULT_MSG_LIMIT = 4;
const uint32 CFE_SB_BUF_MEMORY_BYTES = 524288;
...
const char * const CFE_SB_DEFAULT_ROUTING_FILENAME = "/ram/cfe_sb_route.dat";
...
}}}

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-09-08 12:46:23:_

Replying to [comment:21 cdknight]:
> 
> A simple example:
> 
> cfe_platform_cfg.h:
> {{{
> ...
> extern const uint32 CFE_SB_DEFAULT_MSG_LIMIT;
> extern const uint32 CFE_SB_BUF_MEMORY_BYTES;
> ...
> extern const char * const CFE_SB_DEFAULT_ROUTING_FILENAME;
> ...
> }}}

Actually this is my preferred method storing tunable configuration items and there is already an existing means to do (essentially) this approach.  To minimize impact on the symbol table, a structure is used (rather than an array) so each element does not need to be the same type.  This is currently instantiated by `target_config.c` in the CMake build.

The nice thing about this is that it defers resolution of the value to link time (rather than preprocessor) so code is more modular.  It is particularly useful for values that are used by a different module than what defined them, e.g. defined by CFE and used by CFS apps, or defined by PSP and used by CFE, etc.

The biggest issue with this is that it is not possible to declare derived types from them. e.g.
{{{
uint8 MyBuffer[CFE_SB_BUF_MEMORY_BYTES];
}}}
will break.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-10-06 12:33:09:_

== UPDATED ==

This ticket was last discussed during the 2015-09-08 CCB meeting, and at that time it was requested that the message-handling changes be further separated from the macro name changes.

Ticket #144 was created to address the message-handling changes, namely removing the explicit references to size macros e.g. `OS_MAX_API_NAME` where data is copied to/from message buffers and replacing this with `sizeof()`.  

The above ticket should be considered a prerequisite to this one, as it allows the message payload to be defined in any way, as long as the string is an array (a requirement for SB payloads), and the runtime code will still work.

Branch `trac-12-platform-msg-limits-take2` has been rebased on top of #144 to reflect this.  Commit [changeset:25f11c8] reflects the current proposed update under this ticket, which now consists solely of the macro name changes.

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-11-08 14:17:08:_

Current crop of cfe-next are all going into CFE 6.6

**Comment by skliper** [comment]
_Trac comment by jphickey on 2017-04-27 13:57:46:_

== Resurrecting this changeset for CFE 6.6... ==

Fixing this remaining issue will be required before any EDS changes can be brought in.  In addition this causes problems when attempting to build binaries that are not CPU-specific, since commands and telemetry definitions may vary between CPUs

I have pushed a new branch up to babelfish, called `trac-12-platform-msg-limits-take3`, which is based on today's development branch (2017-04-27).

On this branch there are actually two commits:
[changeset:81c64e0] : rebases the previous 2015-era changeset to the latest development.
[changeset:5099bc5] : fixes up the remaining implicit inclusion of cfe_msgids.h

I kept these as separate commits as they are probably worthy of separate review, and not necessarily tied to one another.

**Comment by skliper** [comment]
_Trac comment by sstrege on 2017-05-16 12:42:52:_

Recommend removing the CFE_PLATFORM_ES_PERF_MAX_IDS configuration macro.  Performance IDs should be a mission wide assignment.  Providing a platform configuration for setting the max is misleading.

Other than that I recommend/approve the changesets.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2017-05-16 14:45:04:_

Commit [changeset:4dd9d3be] addresses the topics discussed in the 2017-05-16 CCB meeting.

- removal of OSAL config includes from message definitions `cfe_sb_msg.h`.  Also removed `osapi.h` from `cfe_evs_msg.h` for the same reason.
- removal of `CFE_PLATFORM_ES_PERF_MAX_IDS` configuration macro and keeping only the mission scope version, to be clear that perf IDS are mission-wide allocations.

The second bullet item required touching several additional files to change references from `CFE_PLATFORM_ES_PERF_MAX_IDS` to `CFE_MISSION_ES_PERF_MAX_IDS`

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-03-05 14:58:28:_

Milestone renamed
